Add "Pipeline Coaster" ride type

X123M3-256 opened the pull request:
This PR adds a new ride type "Pipeline Coaster" which is an Arrow pipeline coaster. This ride type never made it past the prototype stage, but it was featured in RCT3. This ride type gets all sloped and banked pieces, plus inline twists, half loops, medium half loops, and quarter loops. Large corkscrews and vertical twists are available with cheats only.

It does permit vertical slopes although there is no evidence the real thing would have had them. I quite like this because it both permits the use of the quarter loop that is quite useful for this ride, and also makes the ride a bit more different from the corkscrew coaster, and it is consistent with the RCT3 version that also allowed vertical slopes. I decided to make the large corkscrews cheat only though, mostly because I think it just makes the ride type too flexible and more of just a weird looking twister coaster if you have those.

![pipelinepr](https://github.com/user-attachments/assets/c54b7fca-3357-4ade-aa8f-f228653a18c5)

X123M3-256 committed:
Add "Pipeline Coaster" ride type

X123M3-256 committed:
Fix glitch

X123M3-256 commented:
> @X123M3-256 which pieces do we need to look at? The station is one of them, but what is the other one?

The diagonal large flat to steep/steep to flat is the only piece left that is known to glitch (and it glitches badly from several angles). There's probably other glitches I have missed.

The station doesn't glitch, but the way I implemented it won't work properly if you hide the platform (I had baked the platform into the track sprite to make it not glitch) so I need a cleaner way to implement it.

I will upload the vehicle tomorrow.

SpartanFrederic104 commented:
Currently, there isn't an available description for the pipeline coaster. My idea for the description would be something this:

> **_A prototype looping roller coaster where the trains move between the rails_**

SpartanFrederic104 commented:
That's a better description.

X123M3-256 commented:
@Gymnasiast Here is a link to the (unfinished) pipeline coaster train for testing https://mega.nz/file/G5ZX1ACJ#6SChbC6Uq5rCNb9krTs34HjYuxTU13b-1u1lWdzIas0

X123M3-256 committed:
Fix diagonal large flat to steep pieces

X123M3-256 committed:
Fix station drawing

X123M3-256 committed:
Fix sloped turns

X123M3-256 committed:
Add description

X123M3-256 committed:
Add block brake sprites

X123M3-256 commented:
OK, I have fixed every outstanding issue that I knew about so this is now ready for testing.

In order to make the stations work properly I have had to alter the station paint function and add a new station base type, replacing the previous approach that had it baked into the track sprite, but it means that it will now work properly if you hide the station platform (and, in theory, any other ride type that needs such a tall station could use the same station base sprite).

Gymnasiast commented:
@ZeeMaji Could you test this?
(I’ll do some tests of my own if I have time before my holiday.)

X123M3-256 committed:
Fix cheat pieces being available by default

brisingr25 commented:
I've played for a bit, and it's a great coaster type, super fun to play. So far I've only encountered these minor issues:

The track in the building preview for diagonal gentle to flat and flat to gentle pieces appears longer than it actually is:

![diagonal](https://github.com/user-attachments/assets/7046a174-a014-469e-938d-a964ecc16e2d)

A glitch in banked diagonal slopes from two angles only, I don't know if this has been fixed already:

![glitch 2](https://github.com/user-attachments/assets/d05584a2-6a2b-48c0-aa61-db760405b6a1)

And this glitch is a bit of a rare case, but you can build large diagonal slopes next to a medium loop like this, and due to the size of the track, it glitches through the other track:

![glitch](https://github.com/user-attachments/assets/817268d9-bef6-4d84-bcb3-f9f7940cafed)

I don't know if this one is fixable or if anything should be done about it, it's like the helixes glitching in many coaster types. Anyway, I'll play some more when I can and report any issues I find. Great work X123M3-256, as always!

X123M3-256 committed:
Fix oversized diagonal flat to gentle sprite

X123M3-256 committed:
Fix glitch on large banked sloped turns

ZeeMaji commented:
> ZeeMaji Could you test this? (I’ll do some tests of my own if I have time before my holiday.)

Sure, i've already been testing it a little bit thus far on my own terms but i'll shift into full testing now.

X123M3-256 committed:
Fix more glitches

ZeeMaji commented:
On diagonal banked turns, these supports don't connect very well. (Bottom left, top right)
![ksnip_20250706-165343 webp](https://github.com/user-attachments/assets/907068a9-39e6-41d0-bada-6733ab5530be)

brisingr25 commented:
I was about to point out the same thing as ZeeMaji with the banked diagonal turns:

![Captura de pantalla 2025-07-06 232348](https://github.com/user-attachments/assets/9a8cff53-4ab4-46f3-a3b6-3c8153a9d8cf)

They also glitch with the ground or water when built at ground level, but not when the train passes:

![glitch4](https://github.com/user-attachments/assets/e63d2544-ec33-4b0a-bdb4-0f19c718d602)

Also, is it necessary for supports to draw here like the ones on the left? I think other coaster types don't have supports when building turns at ground level, and maybe it would look better:

![Captura de pantalla 2025-07-06 232304](https://github.com/user-attachments/assets/1cd32b07-472c-48ab-87e2-4893918f5b1e)

There's also this wonky thing that happens when a ground level turn support encounters a terrain slope:

![glitch5](https://github.com/user-attachments/assets/b657ea5f-db9f-468a-8881-f2cbc2347326)

Lastly, I don't know if this could be improved, the spine connection of these two pieces (I realise that it's this way because it connects perfectly with other diagonal steep pieces, though):

![Captura de pantalla 2025-07-06 232513](https://github.com/user-attachments/assets/5a3b4bdf-e7d1-4668-9e00-689b234c0ae1)

X123M3-256 committed:
Fix large turns glitching through ground

X123M3-256 commented:
> On diagonal banked turns, these supports don't connect very well

That is very concerning, that bug isn't present in this earlier screenshot. I am worried that something might have gone very badly wrong here, if something was messed up in my support code before the final export then there might be a lot of other supports broken and fixing them all would be an enormous pain at this stage. I need to check for other misaligned supports.

![pipelineturns](https://github.com/user-attachments/assets/49d2fbe8-b0d5-4782-b873-bbf3b587831f)

brisingr25 commented:
Caught another one. Supports for large corkscrews draw without footers when built on ground level:

![glitch6](https://github.com/user-attachments/assets/62041de3-f12a-4f8a-a3a5-d05b0959c45a)

X123M3-256 committed:
Fixed large turn supports

ZeeMaji commented:
With this setup this bit of the track flickers infront of the station platforms.
![ksnip_20250707-214143 webp](https://github.com/user-attachments/assets/8f474d14-8138-4adc-85d8-2fd4f5c7b3ab)

It happens with the ride set to twister too, although to a lesser degree. So i'm unsure if it's entirely specific to the pipeline.
![ksnip_20250707-214259 webp](https://github.com/user-attachments/assets/1439edc1-fcd9-4eaa-bc04-eec8a8f192c5)

Gymnasiast commented:
> And there's also something that isn't unique to the pipeline, so if it's preferable to open a new issue for it let me know: I feel like ride height markings are missing on the new diagonal large flat to steep and steep to flat pieces:

@brisingr25 Yes, please open a new issue for that.

X123M3-256 committed:
Fix vehicle glitching

X123M3-256 committed:
Overwrite heartline twister bounding boxes

X123M3-256 commented:
> Some more small glitching on diagonal slopes, this case unbanked and with the back car of the trains. 

OK, I have fixed this issue but I've also made another change that might be unwanted, I have overwritten the bounding boxes for the heartline twister coaster vehicles with my own. The front car of the pipeline coaster uses a bounding box that is the same size as the heartline twister coaster, and the only reason I did not just reuse the heartline twister coaster bounding boxes for the pipeline coaster is that they were glitching on the pipeline track. This is in large part because the heartline twister coaster does not have correct bounding boxes for turns, it seems Sawyer didn't bother to fill them in correctly, as the heartline twister doesn't have them.

However, having two sets of bounding boxes that are the same size except one is just more glitchy seems a bit redundant - not only is there no reason I can see for any new pipeline car that may be created to use the old set, there also doesn't seem to be any issue caused by having the old cars use the new set. This change has an additional benefit in that you can now put the vanilla heartline twister coaster cars on the pipeline coaster track without them glitching. This may be useful for those who want to use the pipeline pieces as a stand in for pieces that the heartline track is missing, such as verticals and half loops, and I think it just makes sense to have the pipeline vehicles be interchangeable with each other's track like other ride vehicles would be.

This does, however, mean this PR would now alter bounding boxes for the vanilla heartline twister cars. I have tested the vanilla heartline twister with the new bounding boxes and see no issues at all, but I don't know if a change to a vanilla ride is welcome.

Because I changed the vehicle bounding boxes you'll need to download an updated test vehicle, the old one will now have wrong bounding boxes https://mega.nz/file/L8QwUDrB#aEFw7alurJrteVciihh4oEPm9bqqmb7uyiVt3VL-sxU

brisingr25 commented:
Before I continue testing with the new trains, I've checked the tunnels. The twisted vertical pieces don't have tunnel entrances:

![verticaltwist](https://github.com/user-attachments/assets/18df0b15-ca24-43d1-a5f2-35192380d5cd)

This is something I never do when playing because it always looks bad, but if tunnels are built with the lowest clearance the track glitches in some cases. I don't know if this is something to be fixed, the track is larger and taller than other track types so it's to be expected.

![tunnel0](https://github.com/user-attachments/assets/118e53be-bb7e-45ee-b735-e8519a47a6ae)

![tunnel1](https://github.com/user-attachments/assets/b5076bd2-1bd3-4d4b-8d7c-833a295ff1e6)

![tunnel3](https://github.com/user-attachments/assets/176fdcdb-e55a-41a1-bfe5-6817b0bc0c21)

Large corkscrews and quarter loops also glitch in this configuration. Quarter loops tunnel entrances in particular look bad when coming from an inline twist, but I don't think anything can be done about it. Maybe if tunnel entrances were taller like with the inverted coaster. But quarter loops and tunnels and low clearances look bad or glitch in most coaster types:

![tunnel2](https://github.com/user-attachments/assets/0940bedb-1fb8-4977-988b-ecebb05d7bf4)

A lot of this stuff with low clearances has always been like this in vanilla, I assume most people don't build this way because it looks bad.

karst commented:
This is I believe a bounding box issue with the land piece, rather than the track pieces.

X123M3-256 commented:
@mixiate I noticed that tree bug, it was causing a ton of glitching on my test park. I have actually patched it on my local branch because of it was making it hard to spot the glitches with the actual ride type and ruining my GIFs. I didn't commit that patch though because it's unrelated to the pipeline coaster ride type. I might submit a separate PR for it, it's a simple change but I wanted to test it and see that it's not causing other issues.

mixiate commented:
I saw that you had a commit for changing tree bounding boxes. I'm very wary of changing scenery bbs because it affects so much. In the past I tried to change them, and fixed trees glitching like this, but it ended up making everything else worse.

Ultimately it affects everything anyway, not just trees. It's a major flaw/over optimization of the sort algo. If you want to test it fixed, delete these lines:

https://github.com/OpenRCT2/OpenRCT2/blob/65b5d86e6ccc5ad371a2fb826b277faf10b1303d/src/openrct2/paint/Paint.cpp#L478-L481

X123M3-256 committed:
Improve diagonal pieces

brisingr25 commented:
Did some testing with the newest version and the new trains (which look better for sure). Found more instances of footers not drawing when building at ground level:

![supports](https://github.com/user-attachments/assets/6650bba6-4dbd-47d1-9959-309f88c01c0d)

The gap in this special turn piece is still present:

![support](https://github.com/user-attachments/assets/2dac9f14-0ebc-4a0d-9c65-471411c4e230)

In terms of glitches, I've only seen these on diagonal slopes, only from this angle:

![glitch9](https://github.com/user-attachments/assets/a8febbd7-5773-48aa-92d5-309296d8d38d)

X123M3-256 commented:
@brisingr25 The issue with the footers is simply that supports don't draw footers if they're too short. I can certainly fix some of these by increasing the height of the support a bit, but if you go too far it starts poking through the track. I'll see what I can do.

Gymnasiast commented:
I think it’s good to compare those with an original type like the Corkscrew - if that doesn’t draw a footer in a specific situation, it’s okay if the pipeline doesn’t do so either.

X123M3-256 commented:
@Gymnasiast The thing is that the pipeline track is very tall. So the supports end up shorter than for most vanilla tracks. Most vanilla tracks will draw a footer in the locations where the pipeline isn't doing that but that's because the support is taller.

In most cases I think I can force it draw a footer if desired but the result may end up looking worse if the support is very short, at least in my opinion. Some of these supports are only a few pixels, the footer itself is taller than that so it ends up clipping or looking squashed.

Ideally, some of these super short supports just wouldn't draw at all but I'll have to look at if I can make it do that without messing up supports that *aren't* at ground level.

X123M3-256 committed:
Fix some supports not drawing footers

brisingr25 commented:
Yes, I think that for every instance of footers not drawing I checked by converting the pipeline into other coaster types to see if they draw a footer, and they do. But now with X7s' explanation I see why it happens and why it may not be fixable.

X123M3-256 commented:
OK, so I've looked into it further, and I think, the only way to make these super short supports not draw at all is if #23938 was merged but it looks like that was rejected. I'd agree with @brisingr25's original comment that track pieces built at ground level should just not draw a support, but the only way to do that is to set the height to zero and use the extra height parameter, which will make the support draw with a cap and that looks much worse. The pipeline is different from many vanilla ride types because there are a lot more track pieces where I need the support to draw in front of the track rather than behind, owing again to the size of it. Having the supports in the center of the tile as they usually would be means they won't line up with banked track, so all banked turn pieces have the supports on the tile edges where they'd normally be centered.

X123M3-256 committed:
Fix tunnels

mixiate commented:
I can definitely revive #23938, it was partially made with this in mind of course. When I suggested to close I didn't realise that this was right around the corner.

X123M3-256 committed:
Fix more glitches

X123M3-256 commented:
https://private-user-images.githubusercontent.com/146945058/462909278-817268d9-bef6-4d84-bcb3-f9f7940cafed.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIxMDIyMjgsIm5iZiI6MTc1MjEwMTkyOCwicGF0aCI6Ii8xNDY5NDUwNTgvNDYyOTA5Mjc4LTgxNzI2OGQ5LWJlZjYtNGQ4NC1iY2IzLWY5Zjc5NDBjYWZlZC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcwOVQyMjU4NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNzkzNTAyNTVjMmUzYzg2MzA2MDhhM2UwNzBlNjVkYjhhMGZiNWJkNGUxZGI1MDhiNzYyYmRlYTI5MWJkNjcyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jCNuurccjhaOZnIfYNxzoHa8hIFp_D_84fGf9JXCDXw

https://private-user-images.githubusercontent.com/146945058/463424767-abef090f-d704-41a0-a30c-1861f96d275a.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIxMDIyMjgsIm5iZiI6MTc1MjEwMTkyOCwicGF0aCI6Ii8xNDY5NDUwNTgvNDYzNDI0NzY3LWFiZWYwOTBmLWQ3MDQtNDFhMC1hMzBjLTE4NjFmOTZkMjc1YS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcwOVQyMjU4NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lOTNlNmQ2ZjhlNDk1NTY5MWFhYjU1NTExYTY3YzFkZWY1YjgyOWFmMzA0Mjk4ZDY0YjlkYWRjMTMwZjdjNjZkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Vsxdd5chdTW9Yn92ng7Pvv7spo-0xRYnb2DDEb7weiI

https://private-user-images.githubusercontent.com/42477864/463462810-1439edc1-fcd9-4eaa-bc04-eec8a8f192c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIxMDIyMjgsIm5iZiI6MTc1MjEwMTkyOCwicGF0aCI6Ii80MjQ3Nzg2NC80NjM0NjI4MTAtMTQzOWVkYzEtZmNkOS00ZWFhLWJjMDQtZWVjOGE4ZjE5MmM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzA5VDIyNTg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiYjQ3ODdhN2RlZWQzMTUwNTgwYTRjZTM3OTU2ODAzZjhjYWVjOWFhYWI2ZmU2OGFiNWJiMGRmM2JmMmM2YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.sySwW6g2kYF7576tGihd5xgdNNo606Ob0nlRvp1Tf0s

https://private-user-images.githubusercontent.com/146945058/463681180-b5076bd2-1bd3-4d4b-8d7c-833a295ff1e6.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIxMDIyMjgsIm5iZiI6MTc1MjEwMTkyOCwicGF0aCI6Ii8xNDY5NDUwNTgvNDYzNjgxMTgwLWI1MDc2YmQyLTFiZDMtNGQ0Yi04ZDdjLTgzM2EyOTVmZjFlNi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcwOVQyMjU4NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMzJhN2ZiOTI5NjYwN2JmYzI3ODMyZGIzZWYyMzllODgzYzZmOWRiOGEwMDE3NTY0ZTMwMzJiNzc5YzUxOTM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.mAskKJLabu-mHOf3NB-QQL1VZ785ruCqltsH9mlM1sM

X123M3-256 commented:
OK, so to make things clearer I am reposting the issues which have *not* so far been addressed so they're easier to see amongst those that have been fixed already. Some of them I have looked at already but failed to fix.

These medium loops have not been looked at yet. I might be able to fix this particular issue but to be honest, I expect there are many more just like it, and it's impossible to exhaustively test all the different ways track might be placed close to each other. Because this track is wider than most, the standard clearances are a bit too lenient for it and you can sometimes place track in ways where they do actually intersect in 3D space (the go kart track had similar issues with you being able to place things that intersect the track, there's not really a way to avoid clipping in that situation, and not really practical way to give the pipeline its own distinct clearance boxes)

![unaddressed1](https://github.com/user-attachments/assets/1f3a765f-564c-4d4a-a685-30954e849cb9)

This glitch, I tried to fix, it should've been simple but nothing I've tried works and I'm out of ideas

![unaddressed2](https://github.com/user-attachments/assets/4eba0a68-401a-43a4-868e-cd324b9dbeaa)

This is another one I've looked at but to be honest I'm at a loss for what to even try because I've no idea what causes it. The glitch only happens when the train passes over, and not even near the sprite that is glitching, and I can't figure out why it does that. I thought the stable sort might help but it doesn't. I tried adjusting the bounding boxes a bit but it didn't seem to help any.

![unaddressed3](https://github.com/user-attachments/assets/bd46c17f-e47f-4a95-a8b0-46e2b6dd968d)

This one, and the others involving tunnels with low ceilings, has not yet been looked at. This used to be much more severe and the track already underwent a major rewrite prior to submission (which is what stalled it earlier in the year), which has greatly cut down on this from what it was but the issue is, the bounding boxes all need to fit under that low ceiling if I want it to draw right, and sometimes, they need to be higher than that to make it so it won't glitch with the train. Again, the root cause is really the clearances being too lenient - it probably shouldn't allow such a low ceiling. I had considered using a higher clearance height for this ride, as the invert does, but I decided against it both because no non-inverted ride type does that (the heartline twister coaster doesn't), but perhaps more importantly because it makes helixes unbuildable and I wanted to have that option. I can take a look at which of these glitches I can fix without breaking something else but I can't promise all of them.

![unaddressed4](https://github.com/user-attachments/assets/8b528044-8cba-4402-8bf8-8d47f1e2a68f)

X123M3-256 committed:
Fix diagonal gentle to steep glitch

X123M3-256 committed:
Fix track pieces glitching through land on tunnels

X123M3-256 committed:
Fix another broken tunnel

X123M3-256 committed:
Smooth diagonal steep pieces

X123M3-256 commented:
I have managed to address the issue with diagonals not connecting smoothly by manually editing the sprites ... it's not perfect, but it's a lot better than it was.

<img width="216" height="288" alt="diagonals" src="https://github.com/user-attachments/assets/cd680e12-99e6-4735-9f8b-dff88098fa46" />

X123M3-256 commented:
OK, so, I've had a look at the medium loops and I think that's a clear case of the two actually intersecting. The clearance boxes for the medium loops are tight, and were originally set up with the single rail track in mind - with such a bulky track it actually protrudes quite a way into the tile behind. This is made even more clear with two back to back medium loops

<img width="428" height="441" alt="twomedloops" src="https://github.com/user-attachments/assets/dce4f226-0953-487b-84b7-5f78395264f0" />

I don't know what is best done about this. The real fix would be to change the medium loop footprint to add an extra tile, but that would entail adding a special medium loop track piece for just the pipeline coaster. As well as duplicating track pieces this would mean that using arbitrary ride types would not work as expected because there'd essentially be two versions of the same track piece.

I will see if I can adjust the sprites any to make it look a little (because, even though they do overlap in 3D space the sprites could draw in an order that makes it look like they don't) but that might be difficult, it was really hard to get the medium loops to draw correctly with the train, so I'm not sure how much I can change the bounding boxes without breaking something else. But aside from this. I also consider the flickering on the station to be out of scope for this PR because, as @mixiate pointed out, it's an issue with the drawing algorithm and having tried to fix it, I don't think it's going to be possible without fixing the broken sort.

So, I'll take a closer look at the medium loops tomorrow and see if there's anything I can do about them, but aside from that I believe I have resolved every issue that has been found so far and it's ready for further testing.

X123M3-256 committed:
Improve medium half loops slightly

X123M3-256 committed:
Fix minor vehicle glitching on gentle slopes

brisingr25 commented:
After some testing with the new trains and updates, I've found some minor issues.

These weird supports are still present. I've checked and other coaster types don't draw supports in this case, but the pipeline does:

![support3](https://github.com/user-attachments/assets/a9dc0a2e-ccca-43fd-a1fa-42893bf4c8e7)


Can something be done about these support connections to track? It's not bad, but maybe it could be improved. I suppose it'd have to be done manually, though.

<img width="786" height="1081" alt="supports2" src="https://github.com/user-attachments/assets/d5ba2310-a0a4-4e16-9706-e725d29942ac" />



From one angle this happens on the diagonal steep to gentle piece:

![support4](https://github.com/user-attachments/assets/a3abcfa8-998b-4819-bef1-f9e72de22c71)


The back of the train glitches from this angle on gentle to steep diagonal. I realise this would be easier to see in a video...

![glitch11](https://github.com/user-attachments/assets/a4c00ad3-bd03-4aed-99ff-b3982265a01f)


This glitch on the gentle to steep diagonal piece still happens, though now it's even more blink and you'll miss it than before, I had a hard time pausing the game to capture this image, so maybe we should just forget about it:

<img width="531" height="689" alt="glitch14" src="https://github.com/user-attachments/assets/6ca1bfa7-0d07-4c25-9044-3fe9fe00d521" />


Another blink and you'll miss it glitch, only from this angle:

<img width="1000" height="638" alt="glitch15" src="https://github.com/user-attachments/assets/8c209275-5b5b-46e7-b148-a187459c9098" />


And another:

<img width="1589" height="885" alt="glitch13" src="https://github.com/user-attachments/assets/9b646106-c127-4e9a-9c51-1c86bae5d6b5" />


A glitch on quarter loops:

![glitch12](https://github.com/user-attachments/assets/7730afd9-219f-4c8a-a08f-0e0a06c095fe)


With the track being so tall, many pieces of the track clip with the ground when built on ground level, and when the train passes they temporarily "get fixed". This happens with other track types like the B&M, but it's not as pronounced because that track is not as large as the pipeline track, so I don't know if this is something to fix.

![glitchground](https://github.com/user-attachments/assets/a39dceab-dfca-4be2-a3c5-a7abfcfd73ee)


A few things on the special small sloped turn piece. The gap on the support is still there (red circle), a glitch on the train only from this angle (yellow), and the support clips through the track (blue), but from other angles it doesn't have a footer:

<img width="1048" height="1245" alt="special1" src="https://github.com/user-attachments/assets/71f5a5e7-c195-4159-8d88-55c5d2dcbfbe" />

In blue, the same support now without a footer, in red, the track clipping with the ground, like I've explained before:

<img width="1000" height="579" alt="special2" src="https://github.com/user-attachments/assets/0cdabb34-0def-4c00-9f2f-91af4a6f34a9" />


Again, most of these are very minor issues and some could be left in, you see stuff like this all the time in the game if you start nitpicking... So fix what you can.

X123M3-256 committed:
Fix glitch on diagonal gentle to steep

X123M3-256 committed:
Fix support on diagonal gentle to steep

X123M3-256 commented:
> Can something be done about these support connections to track? It's not bad, but maybe it could be improved. I suppose it'd have to be done manually, though.

Are you saying you want them moved up or down? They look right to me.

TheClaudeQC commented:
> This pull request has been marked as stale and will be closed in 14 days if no action is taken. To keep it open, leave a comment or remove the `stale-pr` label. If you're awaiting feedback from a developer, please send us a reminder (either here or on [Discord](https://discord.gg/ZXZd8D8)).

Just saying something to not close it

X123M3-256 commented:
I am unable to reproduce this 

https://private-user-images.githubusercontent.com/146945058/465515794-9b646106-c127-4e9a-9c51-1c86bae5d6b5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg5MTc3MTYsIm5iZiI6MTc1ODkxNzQxNiwicGF0aCI6Ii8xNDY5NDUwNTgvNDY1NTE1Nzk0LTliNjQ2MTA2LWMxMjctNGU5YS05YzUxLTFjODZiYWU1ZDZiNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkyNlQyMDEwMTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMzE2YTFhOTE3YjRhYjMyNDkwNDYyYzkyZjMxNjAyOTc2ZjgyN2ZkYmFiODZiMjEyMWVkYWJlZDIwNWMwZGM4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.AT9PdEcJGpuIO_whcRwloiCJXTbs-hS709hDhLJmLkM

X123M3-256 committed:
More fixes

ZenoArrows commented:
> The top support is drawn behind the track, that's why it looks better. Most normal rides draw most supports behind the track but the pipeline has quite a lot that can't be.

Which track segments can't have the supports drawn behind them? All the screenshots in the PR would look fine with all supports drawn behind the track I think.

karst commented:
I assume no, but I am just checking up, has any more progress been made on fixing the glitching since last push?

X123M3-256 commented:
> I assume no, but I am just checking up, has any more progress been made on fixing the glitching since last push?

No sorry. I know it would need just a few more days of work if I could actually work on it but I am just completely gone, every waking moment is agony. I don't know if I'll be able to finish it and I'm wondering if I should just close the PR.

DanMacK1977 commented:
What needs to be completed on it?  Coding or spritework?  I'd be interested in trying any fixes, I'd just have to get a test version up and running.

X123M3-256 commented:
> What needs to be completed on it? Coding or spritework? I'd be interested in trying any fixes, I'd just have to get a test version up and running.

A little bit of both, kind of. The remaining issues are visual glitches where sprites are drawn in the wrong order, like parts of the track being drawn behind the train when they should be in front, etc.

Each track piece has to be split into multiple sprites for drawing. The spritework is mostly done in the sense that no new sprites need to be drawn, but it is necessary to edit some of the existing sprites by moving parts of the track piece from one sprite to another. This is one way that these draw order issues can be addressed.

The other way to fix it is by editing the bounding box for the sprite. Each sprite has a bounding box which essentially determines its position in 3D space as far as the drawing engine is concerned. Sometimes, editing this box (which is a change to the code) is the easier way to make things draw in the right order. And often you do have a choice (a given glitch might be able to be fixed with a change to the bounding box *or* a change to the sprites).

But there is, at this point, no modelling or rendering work left to complete and no real programming beyond fiddling with those bounding box values. It just needs the remaining glitches fixed so it draws properly.

DanMacK1977 commented:
Okay, I'll take a look at it and see what I can contribute as well.

DanMacK1977 commented:
Playing with this, I find a few minor issues, mainly with track glitching (which happens on alot of vanilla rides to some extent), and the support drawing issue, which apparently is somewhat difficult to deal with and is moreso a coding limitation issue than a spriting issue. 

<img width="981" height="730" alt="MedLoopGlitch1" src="https://github.com/user-attachments/assets/fdc8e45b-7595-4840-a0f2-41c17e0efd49" />
Bit of a glitch on the facing side of a medium loop (Sorting glitch or sprite fiddling?)
<img width="957" height="727" alt="Pipeline Support Glitch1" src="https://github.com/user-attachments/assets/2b7b3625-7dec-4732-9576-db50c0222028" />
Support drawn through track (Might be unavoidable)
<img width="567" height="573" alt="SmallFlatTurnGlitch" src="https://github.com/user-attachments/assets/4c2a4f80-ffbe-4326-80c3-4200bd111279" />

Bit of a glitch on the small flat turn.

If these are pixel pushing issues, I'm happy to fiddle with that.  It might almost be worth putting it in with the glitches, most of them are blink and you'll miss them, and the support issue, as X123M3 said, might be unavoidable.  The clipping on the back to back Medium Loops is also unavoidable, but I think casual players won't mind and more conscious players will add a section between them.  

Graphical errors and glitches from Vanilla are still being fixed 20+ years later, but I think this ride deserves some more widespread testing.  Think of it as Early Access?  It's too fun of a coaster not to be in trunk.  

That's just my 0.02, but it's from a very long-time player.   Shame for it to sit in limbo like the prototype, lol.

DanM

Realsteel89 commented:
Think the support in that curve you can't fix because the track using the full tile as size thats why the support glitch
